community/lucene-index/src/main/java/org/neo4j/kernel/api/impl/index/collector/DocValuesCollector.java

```diff
@@ -312,6 +320,92 @@ private void replayTo( Collector collector ) throws IOException
         }
     }
 
+    /**
+     * Iterates over all per-segment {@link DocValuesCollector.MatchingDocs}.
+     * Provides base functionality for extracting entity ids and other values from documents.
+     */
+    private abstract static class LongValuesSource
+    {
+        private final Iterator<DocValuesCollector.MatchingDocs> matchingDocs;
+        private final String field;
+        final int totalHits;
+        final Map<String,NumericDocValues> docValuesCache;
+
+        DocIdSetIterator currentIdIterator;
+        NumericDocValues currentDocValues;
+        DocValuesCollector.MatchingDocs currentDocs;
+        int index;
+        long next;
+
+        LongValuesSource( Iterable<DocValuesCollector.MatchingDocs> allMatchingDocs, int totalHits, String field )
+        {
+            this.totalHits = totalHits;
+            this.field = field;
+            matchingDocs = allMatchingDocs.iterator();
+            docValuesCache = new HashMap<>();
+        }
+
+        /**
+         * @return true if it was able to make sure, that currentDisi is valid
+         */
+        boolean ensureValidDisi()
+        {
+            try
+            {
+                while ( currentIdIterator == null )
+                {
+                    if ( matchingDocs.hasNext() )
+                    {
+                        currentDocs = matchingDocs.next();
+                        currentIdIterator = currentDocs.docIdSet.iterator();
+                        if ( currentIdIterator != null )
+                        {
+                            docValuesCache.clear();
+                            currentDocValues = currentDocs.readDocValues( field );
+                        }
+                    }
+                    else
+                    {
+                        return false;
+                    }
+                }
+                return true;
+            }
+            catch ( IOException e )
+            {
+                throw new RuntimeException( e );
+            }
+        }
+
+        boolean fetchNextEntityId()
+        {
+            try
+            {
+                if ( ensureValidDisi() )
+                {
+                    int nextDoc = currentIdIterator.nextDoc();
+                    if ( nextDoc != DocIdSetIterator.NO_MORE_DOCS )
+                    {
+                        index++;
+                        next = currentDocValues.get( nextDoc );
+                        return true;
+                    }
+                    else
+                    {
+                        currentIdIterator = null;
+                        return fetchNextEntityId();
```

Comment:
Is this recursive call really necessary? If it is, that's fine, but I would prefer something that is easier to reason about.

Reply:
This isn't changed though... so something for another PR/discussion?